packages/validator/src/services/syncCommittee.ts

```diff
@@ -0,0 +1,182 @@
+    // Before altair fork no need to check duties
+    if (computeEpochAtSlot(this.config, slot) < this.config.params.ALTAIR_FORK_EPOCH) {
+      return;
+    }
+
+    // Fetch info first so a potential delay is absorved by the sleep() below
+    const dutiesAtSlot = await this.dutiesService.getDutiesAtSlot(slot);
+    const dutiesBySubcommitteeIndex = groupSyncDutiesBySubCommitteeIndex(this.config, dutiesAtSlot);
+    if (dutiesAtSlot.length === 0) {
+      return;
+    }
+
+    // Lighthouse recommends to always wait to 1/3 of the slot, even if the block comes early
+    await sleep(this.clock.msToSlotFraction(slot, 1 / 3), signal);
+
+    // Step 1. Download, sign and publish an `SyncCommitteeSignature` for each validator.
+    //         Differs from AttestationService, `SyncCommitteeSignature` are equal for all
+    const beaconBlockRoot = await this.produceAndPublishSyncCommittees(slot, dutiesAtSlot);
+
+    // Step 2. If an attestation was produced, make an aggregate.
+    // First, wait until the `aggregation_production_instant` (2/3rds of the way though the slot)
+    await sleep(this.clock.msToSlotFraction(slot, 2 / 3), signal);
+
+    // await for all so if the Beacon node is overloaded it auto-throttles
+    // TODO: This approach is convervative to reduce the node's load, review
+    await Promise.all(
+      Array.from(dutiesBySubcommitteeIndex.entries()).map(async ([subcommitteeIndex, duties]) => {
+        if (duties.length === 0) return;
+        // Then download, sign and publish a `SignedAggregateAndProof` for each
+        // validator that is elected to aggregate for this `slot` and `subcommitteeIndex`.
+        await this.produceAndPublishAggregates(slot, subcommitteeIndex, beaconBlockRoot, duties).catch((e) => {
+          if (notAborted(e))
+            this.logger.error("Error on SyncCommitteeContribution", {slot, index: subcommitteeIndex}, e);
+        });
+      })
+    );
+  };
+
+  /**
+   * Performs the first step of the attesting process: downloading `SyncCommittee` objects,
+   * signing them and returning them to the validator.
+   *
+   * https://github.com/ethereum/eth2.0-specs/blob/v0.12.1/specs/phase0/validator.md#attesting
+   *
+   * Only one `SyncCommittee` is downloaded from the BN. It is then signed by each
+   * validator and the list of individually-signed `SyncCommittee` objects is returned to the BN.
+   */
+  private async produceAndPublishSyncCommittees(slot: Slot, duties: SyncDutyAndProof[]): Promise<Root> {
+    const logCtx = {slot};
+
+    // /eth/v1/beacon/blocks/:blockId/root -> at slot -1
+
+    // Produce one attestation data per slot and subcommitteeIndex
+    // Spec: the validator should prepare a SyncCommitteeSignature for the previous slot (slot - 1)
+    // as soon as they have determined the head block of slot - 1
+    const beaconBlockRoot = await this.apiClient.beacon.blocks.getBlockRoot(slot - 1).catch((e) => {
```

Comment:
my understanding is we want to get block root at the current slot since we wait for 1/3 slot already, so `getBlockRoot(slot)`

correct me if I'm wrong

Reply:
will address in future PRs

Reply:
Yeah that's right, we want to sign the head at that slot since the signatures will be included in the next. Even we can / should do `getBlockRoot("head")`

Reply:
See https://github.com/ChainSafe/lodestar/pull/2522 as conclusion